### ISSUE TITLE
msgp may wish to handle function types gracefully

### ISSUE THREAD

**Issue by glycerine** [tinylib/msgp#127]
When I define a type that is a function, e.g. Filt below, msgp may wish to ignore it, rather than to try to serialize it and choke.

example:
~~~
package main

//go:generate msgp                                                                                                                          

type Filt func(n int) bool

// a collection of Filt                                                                                                                     
type FilterSet struct {
    Map map[string]Filt
}

func main() {}
~~~

example output of generate and build:
~~~
jason@jasonsmac ~/tmp $ go generate
======== MessagePack Code Generator =======
>>> Input: "f.go"...                                                                                                                        
parsing FilterSet...                                                                                                                        
 ⚠ non-local identifier: Filt                                                                                                               
 ⚠ WARNING: unresolved identifier: Filt                                                                                                     
>>> Wrote and formatted "f_gen.go"                                                                                                          
>>> Wrote and formatted "f_gen_test.go"                                                                                                     
>>> Done.                                                                                                                                   
jason@jasonsmac ~/tmp $ ls
f.go        f.go~       f_gen.go    f_gen_test.go
jason@jasonsmac ~/tmp $ go build
# _/Users/jason/tmp
./f_gen.go:48: bzg.DecodeMsg undefined (type Filt has no field or method DecodeMsg)
./f_gen.go:81: bzg.EncodeMsg undefined (type Filt has no field or method EncodeMsg)
./f_gen.go:98: bzg.MarshalMsg undefined (type Filt has no field or method MarshalMsg)
./f_gen.go:143: bzg.UnmarshalMsg undefined (type Filt has no field or method UnmarshalMsg)
./f_gen.go:165: bzg.Msgsize undefined (type Filt has no field or method Msgsize)
jason@jasonsmac ~/tmp $ 
~~~

**Comment by philhofer** [comment]
I agree that it would improve the user experience. Right now, though, if you have a field that you know isn't serializable, the right thing to do is annotate it with ``` `msg:"-"` ```. In general, the code generator assumes the user knows best, and so in this case it assumes (incorrectly, in this case) that `Filt` implements the appropriate methods somewhere else in the source. (Note that your example runs fine if you don't include the `FilterSet` type. It's only because you've included the type in another declaration that it tries to do something with it.)

I'm open to suggestions about how you think this should work.

**Comment by glycerine** [comment]
I agree. The more I think about, the more it makes sense that, as you say, perhaps even function types could have the appropriate methods defined elsewhere. At first I was thinking, "there's no way to ever serialized the function itself, so why even try?" But upon reflection that is wrong. There are plenty of ways in which a function type could have methods defined on it that represent the serialized version of that function. I added `msg:"-"` to my actual code and all is well now. Closing this issue.
